File: algorithms/python/_018_4Sum/_018_4Sum.py
class Solution1:
    """
    @param numbers: Give an array
    @param target: An integer
    @return: Find all unique quadruplets in the array which gives the sum of zero
    """
    def fourSum(self, numbers, target):
        numbers.sort()
        res = []
        for i in range(len(numbers) - 3):
            if i and numbers[i] == numbers[i - 1]:
                continue
            for j in range(i + 1, len(numbers) - 2):
                if j != i + 1 and numbers[j] == numbers[j - 1]:
                    continue
                sum = target - numbers[i] - numbers[j]
                left, right = j + 1, len(numbers) - 1
                while left < right:
                    if numbers[left] + numbers[right] == sum:
                        res.append([numbers[i], numbers[j], numbers[left], numbers[right]])
                        left += 1
                        right -= 1
                        while left < right and numbers[left] == numbers[left - 1]:
                            left += 1
                        while left < right and numbers[right] == numbers[right + 1]:
                            right -= 1
                    elif numbers[left] + numbers[right] > sum:
                        right -= 1
                    else:
                        left += 1
        return res


class Solution2:
    """
    @param numbers: Give an array
    @param target: An integer
    @return: Find all unique quadruplets in the array which gives the sum of zero
    """

    def fourSum(self, nums, target):
        if not nums or len(nums) < 4:
            return []

        nums.sort()
        result = []
        n = len(nums)
        for i in range(0, n - 3):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            for j in range(i + 1, n - 2):
                if j > i + 1 and nums[j] == nums[j - 1]:
                    continue
                sum = target - nums[i] - nums[j]
                l, r = j + 1, n - 1
                while l < r:
                    if nums[l] + nums[r] == sum:
                        result.append([nums[i], nums[j], nums[l], nums[r]])
                        l += 1
                        r -= 1
                        # remove duplicates:
                        while l < r and nums[l] == nums[l - 1]:
                            l += 1
                        while l < r and nums[r] == nums[r + 1]:
                            r -= 1

                    elif nums[l] + nums[r] < sum:
                        l += 1

                    else:
                        r -= 1

        return result

File: algorithms/cpp/_018_4Sum/_018_4Sum.cpp

# include <iostream>
#include <vector>

using namespace std;

int main() {
    cout << "new file!" << endl;
    return 0;
}

class Solution1 {
public:
    /*
    题意：找到数列中所有和等于目标数的四元组，需去重
    多枚举一个数后，参照3Sum的做法，O(N^3)
    */
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        int len = nums.size();
        int left, right, sum;
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        vector<int> tmp;
        for (int i = 0; i < len - 3; i++) {
            if (i && nums[i] == nums[i - 1]) continue;
            for (int j = i + 1; j < len - 2; j++) {
                if (j != i + 1 && nums[j] == nums[j - 1]) continue;
                sum = target - nums[i] - nums[j];
                left = j + 1;
                right = len - 1;
                while (left < right) {
                    if (nums[left] + nums[right] == sum) {
                        tmp.clear();
                        tmp.push_back(nums[i]);
                        tmp.push_back(nums[j]);
                        tmp.push_back(nums[left]);
                        tmp.push_back(nums[right]);
                        res.push_back(tmp);
                        left++;
                        right--;
                        while (left < right && nums[left] == nums[left - 1]) left++;
                        while (left < right && nums[right] == nums[right + 1]) right--;
                    } else
                    if (nums[left] + nums[right] > sum) right--;
                    else left++;
                }
            }
        }
        return res;
    }
};

class Solution2 {
public:
    /**
     * @param numbers: Give an array
     * @param target: An integer
     * @return: Find all unique quadruplets in the array which gives the sum of zero
     */
    vector<vector<int>> fourSum(vector<int> &numbers, int target) {
        // write your code here
        /*
        -2 -1 0 0 1 2
        -1 -1 -1 -1 0 0 1 1 1 1 2
        */
        vector<vector<int>> result;
        std::sort(numbers.begin(), numbers.end());
        for (int i=0; i<numbers.size(); i++) {
            if (i>0 && numbers[i]==numbers[i-1]) continue;
            for (int j=numbers.size()-1; j>=0; j--) {
                if (j<numbers.size()-1 && numbers[j]==numbers[j+1]) continue;
                if ((j-i)<3) break;
                int sum = numbers[i] + numbers[j];
                int k = i+1;
                int l = j-1;
                while (k<l) {
                    int new_sum = sum + numbers[k] + numbers[l];
                    if (new_sum==target) {
                        result.push_back({numbers[i],numbers[k],numbers[l],numbers[j]});
                        k++;
                        l--;
                        while (k<numbers.size()-1 && numbers[k]==numbers[k-1]) k++;
                        while (l>0 && numbers[l]==numbers[l+1]) l--;
                    } else if (new_sum < target) k++;
                    else l--;
                }
            }
        }
        return result;
    }
};